metaserver/src/main/scala/scala/meta/languageserver/Positions.scala

```diff
@@ -0,0 +1,48 @@
+package scala.meta.languageserver
+
+import org.langmeta.inputs.Input
+
+object Positions {
+  def positionToOffset(
+      input: Input.VirtualFile,
+      line: Int,
+      column: Int
+  ): Int =
+    positionToOffset(input.path, input.value, line, column)
+
+  def positionToOffset(
+      filename: String,
+      contents: String,
+      line: Int,
+      column: Int
+  ): Int = {
+    def peek(idx: Int): Int =
+      if (idx < contents.length) contents.charAt(idx) else -1
+    var i, l = 0
+    while (i < contents.length && l < line) {
+      contents.charAt(i) match {
+        case '\r' =>
+          l += 1
+          if (peek(i + 1) == '\n') i += 1
+
+        case '\n' =>
+          l += 1
+
+        case _ =>
+      }
+      i += 1
+    }
+
+    if (l < line)
+      throw new IllegalArgumentException(
+        s"$filename: Can't find position $line:$column in contents of only $l lines long."
+      )
+    if (i + column < contents.length)
+      i + column
+    else
+      throw new IllegalArgumentException(
+        s"$filename: Invalid column. Position $line:$column in line '${contents.slice(i, contents.length).mkString}'"
+      )
+  }
+
+}
```

Comment:
An alternative suggestion in a completely different style.  Nothing wrong with this one though if you want to keep it.  Ignore my lack of filename in the IllegalArgumentExceptions, I just whipped this up in a worksheet.

```scala
def pos(content: String, line: Int, col: Int): Int = {
  val (lineCount, pos) = content
    .linesWithSeparators
    .take(line + 1)
    .foldLeft((0, 0)){ (accum, str) =>
      val (lineNum, charCount) = accum
      if (lineNum < line) {
        (lineNum + 1, charCount + str.length)
      } else {
        require(str.length >= col, "Line is not long enough")
        (lineNum + 1, charCount + col)
      }
    }
  require(lineCount == line + 1, "Not enough lines")
  pos
}
```

Reply:
This method is on a fairly critical path since it's run for every start/end point during completions. `foldLeft` + tuples creates a lot of unnecessary allocations compared to the while loops. 

However, I just found out that scala.meta.Input has an internal `private lazy val cachedLineIndices: Array[Int]` which allows you to get the offset of a line in from an array lookup. I've refactored to use this instead of rolling our own.